### PR TITLE
[ValueTracking] Improve knownbits from conditions where `Val` is used by `and`/`or`.

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[ValueTracking] Add tests for computing knownbits from `(icmp upred X (and/or X, Y))`; NFC**
- **[ValueTracking] compute knownbits from `(icmp upred X (and/or X, Y))`; NFC**

Proofs: https://alive2.llvm.org/ce/z/K4vFn-

### CODE REVIEW

**Comment by dtcxzyw** [comment]
> [ValueTracking] Add basic tests tracking or disjoint conditions as add; NFC
[ValueTracking] Tracking or disjoint conditions as add in Assumption/DomCondition Cache

I would like to file a separate PR for the above commits :)

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (dtcxzyw) [inline llvm/lib/Analysis/ValueTracking.cpp:712]: Could you please provide the alive2 proof?

**Comment by goldsteinn** [comment]
> > [ValueTracking] Add basic tests tracking or disjoint conditions as add; NFC
> > [ValueTracking] Tracking or disjoint conditions as add in Assumption/DomCondition Cache
> 
> I would like to file a separate PR for the above commits :)

Done, see: https://github.com/llvm/llvm-project/pull/86302

Once that gets in ill rebase this one (merge conflicts otherwise).

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:712]: See desc

**Comment by goldsteinn** [comment]
Rebased

**Comment by goldsteinn** [comment]
ping

**Comment by goldsteinn** [comment]
rebased

**Review by nikic — COMMENTED** [review]
The implementation looks ok to me. I'll leave it to @dtcxzyw to decide if we really need this. Diffs don't look particularly compelling to me.

**Comment by goldsteinn** [comment]
> The implementation looks ok to me. I'll leave it to @dtcxzyw to decide if we really need this. Diffs don't look particularly compelling to me.

Yeah I guess an empty diff isn't really all that exciting. For my money, if there is no-compile time impact, its better to have it covered than not, especially since I think there will end up being minimal code complexity added since the `add nuw`/`sub nuw` cases have a positive diff so if those get in, the impl of the two cases will overlap. But LMK either way.

**Review by dtcxzyw — APPROVED** [review]

**Comment by goldsteinn** [comment]
@nikic, going to push this unless you have any objections.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request closed without merging**
